### PR TITLE
DROOLS-2932: Extend KeyboardOperations to support header cells navigation

### PR DESCRIPTION
**Pull request by jomarko**
Dependent on #558 

Ensemble with:
- kiegroup/kie-wb-common#2200
- kiegroup/drools-wb#968

### CODE REVIEW

**Comment by manstis** [comment]
@jomarko would you like me to take a look yet?

**Comment by jomarko** [comment]
@manstis happy if you have a look on code, architecture, I mean if I am moving in a good way, don't waste time with manual check please.

**Comment by manstis** [comment]
@jomarko OK, so my initial observation would be to move the extra logic from the different `KeyboardOperation` to `BaseCellSelectionManager.adjustSelection(..)`. The `KeyboardOperation` calls `CellSelectionManager.adjustSelection(..)` that (via the `GridWidget`) handles selection changes.

**Comment by jomarko** [comment]
Tests added.

**Comment by ederign** [comment]
@manstis could you please also take a look on this one?

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/selections/impl/BaseCellSelectionManager.java [hunk 1]
```diff
@@ -254,6 +337,22 @@ private boolean moveSelection(final GridData.SelectedCell origin,
...
+                                uiColumnIndex,
+                                false,
+                                false);
+    }
+
+    private boolean possibleToSelectHeaderCell() {
+        return !simpleInnerGridIsSelected() && gridModel.getHeaderRowCount() > 0;
+    }
+
+    private boolean simpleInnerGridIsSelected() {
```
Comment (manstis) [inline uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/selections/impl/BaseCellSelectionManager.java]: I think this is DMN specific? What purpose does it serve? Does it relate to `LiteralExpressionGrid` and `UndefinedExpressionGrid`?

**Comment by kiereleaseuser2** [comment]
Build finished. 5024 tests run, 11 skipped, 0 failed.

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/selections/impl/BaseCellSelectionManager.java (hunk 1 above)
Comment (jomarko) [reply uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/selections/impl/BaseCellSelectionManager.java]: Yes, related to those, should I remove it according to you feedback on kie-wb-common of the ensemble?

**Comment by kiereleaseuser2** [comment]
Build finished. No test results found.

**Comment by kiereleaseuser2** [comment]
Build finished. 5024 tests run, 11 skipped, 0 failed.

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/selections/impl/BaseCellSelectionManager.java (hunk 1 above)
Comment (manstis) [reply uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/selections/impl/BaseCellSelectionManager.java]: @jomarko I'd like to understand _what_ this does? It suggests you can **only** move to the header _if_ the current grid is not a `LiteralExpressionGrid` or `UndefinedExpressionGrid`. What about an "top-level `LiteralExpressionGrid`" that has one row, one columns **and** a header cell? I'll try this scenario in a minute.. when I retest the `kie-wb-common` PR.

Comment (jomarko) [reply uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/selections/impl/BaseCellSelectionManager.java]: @manstis yes, you are right **What about an "top-level LiteralExpressionGrid" that has one row, one columns and a header cell?** doesn't work, so will check after you open a PR for not adding header into nested grids.

**Comment by kiereleaseuser2** [comment]
Build finished. 5040 tests run, 11 skipped, 0 failed.

**Comment by kiereleaseuser2** [comment]
Build finished. 5040 tests run, 11 skipped, 0 failed.

**Comment by kiereleaseuser2** [comment]
Build finished. 5042 tests run, 11 skipped, 0 failed.

**Comment by kiereleaseuser2** [comment]
Build finished. 5047 tests run, 11 skipped, 0 failed.

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/test/java/org/uberfire/ext/wires/core/grids/client/widget/grid/selections/impl/BaseCellSelectionManagerTest.java [hunk 2]
```diff
@@ -544,6 +546,118 @@ public void adjustSelectionRightDownWithShiftKey() {
...
+    public void testAdjustSelectionUpFromDataToHeader() {
+        cellSelectionManager.selectCell(0,
+                                        1,
+                                        false,
+                                        false);
+
+        cellSelectionManager.adjustSelection(SelectionExtension.UP, false);
+
+        final List<GridData.SelectedCell> selectedCells = gridWidgetData.getSelectedCells();
+        Assertions.assertThat(selectedCells).isEmpty();
```
Comment (manstis) [inline uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/test/java/org/uberfire/ext/wires/core/grids/client/widget/grid/selections/impl/BaseCellSelectionManagerTest.java]: We'd normally have a static import in this scenario; however it's fine as it is really.

**Comment by kiereleaseuser2** [comment]
Build finished. 5059 tests run, 11 skipped, 0 failed.

**Comment by jomarko** [comment]
@kiegroup/gatekeepers please merge

**Comment by manstis** [comment]
@jomarko We still need @ederign to approve before merging. Sorry.

**Comment by manstis** [comment]
Merged before @ederign gives final approval as @mareknovotny has merged the related PRs. Without this merged the build will be broken... and it looks like @ederign request for tests has been fullfilled.

**Comment by mareknovotny** [comment]
@manstis thanks! I realized later you asked for re-review of this

**Pull request merged**